### ISSUE TITLE
`formatexpr=Fixedqg` silently overrides user set `formatprg`

### ISSUE THREAD

**Issue by cideM** [HerringtonDarkholme/yats.vim#209]
I originally reported this problem in [Neovim](https://github.com/NixOS/nixpkgs/issues/96921) since that's where the runtime file is, but I think it's just taken from `yats` or at least parts of it are.

[`typescript.vim`](https://raw.githubusercontent.com/neovim/neovim/master/runtime/indent/typescript.vim) sets up `formatexpr` which has priority over `formatprg` for the `gq` mapping, as is explained in the Neovim `:h`

```
gq{motion}		Format the lines that {motion} moves over.
			Formatting is done with one of three methods:
			1. If 'formatexpr' is not empty the expression is
			   evaluated.  This can differ for each buffer.
			2. If 'formatprg' is not empty an external program
			   is used.
			3. Otherwise formatting is done internally.
```

The same happens [in `yats.vim`](https://github.com/HerringtonDarkholme/yats.vim/blob/master/indent/typescript.vim#L17-L19), obviously, since that's where the runtime file in Neovim comes from.

I'm reporting this is kind of a bug since this behavior is quite challenging to debug especially for people who are new to Neovim. The thing is that there's nothing warning you about this behavior and at least in my experience `formatexpr` is a pretty rare sight. If you search for "how to format a file in Neovim" you'll most likely come across various plugins, but also recommendations to leverage `formatprg` and `gq`.

Now I was wondering... what's the purpose of the `Fixedgq` function and can this maybe be handled separately? For example, users who want the behavior of that function could just wrap `gq` with it, or bind it to a key as a `<Plug>` mapping, or it could at least be behind a global and/or buffer local option.

**Comment by equt** [comment]
This is exactly the problem I came across these few days. Instead of using Vimscript to rewrite one, I would prefer using `prettier`, the de facto standard formatting tool for javascript or typescript.

So it would lead to the following settings in my `ftplugin/typescript.vim`.

```vim
let s:prettier_config = {
	" ...
\ }

let &l:formatexpr = 'prettier' . utils#map_as_cli_options(s:prettier_config)
```

But all stuff in `after/ftplugin` loads after `after/indent`, so I have to create my `after/indent` for overwriting that.

I would suggest removing the `formatexpr` setting from `indent` or moving it into `ftplugin`. `typescript` seems to be the only language I found to set the `formatexpr` inside the `indent`. [Python], [JavaScript], [OCaml] and [Rust] maintained by the Rust lang team only place indention related settings inside the `indent` directory.

[Python]: https://github.com/neovim/neovim/blob/master/runtime/indent/python.vim
[JavaScript]: https://github.com/neovim/neovim/blob/master/runtime/indent/javascript.vim
[OCaml]: https://github.com/neovim/neovim/blob/master/runtime/indent/ocaml.vim
[Rust]: https://github.com/rust-lang/rust.vim/blob/master/indent/rust.vim

c.c. neovim/neovim#13113

**Comment by equt** [comment]
> I would suggest removing the formatexpr setting from indent or moving it into ftplugin.

Now moved to the `ftplugin`.

> For example, users who want the behavior of that function could just wrap gq with it, or bind it to a key as a <Plug> mapping, or it could at least be behind a global and/or buffer local option.

Instead of adding a `<Plug>` mapping or a global option, I simply add code to check whether the format tool has been set at that moment. Also, it's been set to `formatprg` for a lower priority.

PR #212

**Comment by cideM** [comment]
Wow that was quick, thanks!

**Comment by BenoitZugmeyer** [comment]
👋 It seems that the change brought by PR #212 is breaking `gqq` (or `gq` in visual mode): this command now inserts `zsh:1: unknown file attribute: v` in the buffer instead of formating lines.

See user reports:
* https://superuser.com/questions/1600371/gq-in-vim-resulting-in-shell-error
* https://www.reddit.com/r/vim/comments/jpz6wa/gq_in_vim_resulting_in_shell_error/

**Comment by equt** [comment]
> I don't know about NeoVim, but in Vim, `&formatprg` is an external command, and you can't really expect external commands to have access to Vim's internal variables.

Yes, this should be exactly the point here, and it's hard to believe that I even never notice that before. Sorry for the disruption, @BenoitZugmeyer, and I have fixed that in PR #216.

**Comment by David-Else** [comment]
I am unsure how this repo relates to Neovim official, but I just wasted hours with this horrible bug! Does the fact it is now fixed here mean it will be fixed in Neovim, or does someone need to say something to the Neovim people? Cheers!

**Comment by brianhuster** [comment]
It is not fixed upstream. But I have opened an issue there

**Comment by brianhuster** [comment]
This bug has been fixed in both Vim and Neovim
